092-self-healing-using-cruise-control.md

```diff
@@ -0,0 +1,284 @@
+* FIX_STARTED - The anomaly has started getting fixed
+* FIX_FAILED_TO_START - The proposal generation for the anomaly fix failed 
+* CHECK_WITH_DELAY - The anomaly fix is delayed 
+* LOAD_MONITOR_NOT_READY - Monitor which is monitoring the workload of a Kafka cluster is in loading or bootstrap state
+* COMPLETENESS_NOT_READY - Completeness not met for some goals. The completeness describes the confidence level of the data in the metric sample aggregator. If this is low then some Goal classes may refuse to produce proposals.
+
+We can also poll the `state` endpoint, with `substate` parameter as `executor`, to get details on the tasks that are currently being performed by Cruise Control.
+For example:
+```shell
+ExecutorState: {state: INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS,
+ pending(20)/in-progress(0)/aborting(0)/finished(0)/total(20) inter-broker partition movements, 
+ completed(0)/total(0) bytes in MBs: 0.00%, max/min/avg concurrent inter-broker partition movements per-broker: 5/5/5.00, 
+ triggeredSelfHealingTaskId: 4b1ed191-50d7-4aa1-806b-b8c5c0d679fe, triggeredTaskReason:
+ Self healing for DISK_FAILURE: {
+	Disk /var/lib/kafka/data-1/kafka-log2 on broker 2 failed at 2024-12-09T09:49:46Z
+ }}
+```
+
+### Enabling self-healing configuration in the Strimzi operator
+
+Every configuration related to self-healing is currently disabled in the operator, and we will need to enable users altering some of those properties.
+The setting of these properties will be placed behind a feature gate.
+
+### Feature Gate
+
+The new feature gate will be called `UseSelfHealing`.
+It will be introduced in an alpha state and will be disabled by default.
+
+We are using a feature gate for this feature due to various reasons:
+* We want to allow users to try this feature and test it.
+* We would like to know more about the positive/negatives of this feature by testing it in real deployments.
+* Getting feedback from the community would allow us to understand how we can improve this even more.
+
+### Enabling self-healing in the Kafka custom resource
+
+When the feature gate is enabled, the operator will set the `self.healing.enabled` Cruise Control configuration to `true` and allow the users to set other appropriate `self-healing` related [configurations](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations).
+
+Here is an example of how the configured Kafka custom resource could look:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    config:
+      # configuration to enable self-healing for all anomalies
+      self.healing.enabled: true
+```
+
+Where the user wants to enable self-healing for some specific anomalies, they can use configurations like `self.healing.broker.failure.enabled` for broker failure, `self.healing.goal.violation.enabled` for goal violation etc.
```

Comment:
Is there any anomalies enabled by default? If user does not specify anomalies to detect, self-healing will do nothing?

Reply:
If the user doesn't specify any specific anomalies(i.e you used just `self.healing.enabled: true`) then self healing would be detecting all the anomalies. In case the user knows that somehow any anomaly will keep appearing since they have some issues in their cluster, then in that they can use the above property to enable self healing for only specific anomaly and skip the check for the other anomaly